iquip/apps/logger.py

```diff
@@ -0,0 +1,192 @@
+        levelBox: A comboBox for setting the logger's level.
+    """
+
+    def __init__(self, parent: Optional[QObject] = None):
+        """Extended."""
+        super().__init__(parent=parent)
+        # widgets
+        self.logEdit = QTextEdit(self)
+        self.logEdit.setReadOnly(True)
+        self.clearButton = QPushButton("Clear", self)
+        self.levelBox = QComboBox(self)
+        # layout
+        layout = QVBoxLayout(self)
+        layout.addWidget(self.logEdit)
+        layout.addWidget(self.clearButton)
+        layout.addWidget(self.levelBox)
+
+
+class ConfirmClearingFrame(QWidget):
+    """A confirmation frame for log clearing in the LoggerFrame.
+    
+    Attributes:
+        label: The label for displaying a confirmation message to clear logs in the LoggerFrame.
+        buttonBox: The buttonBox with OK and Cancel button to check whether to clear logs.
+        confirmed: A pyqtSignal that emits signal when Ok button clicked. 
+    """
+
+    confirmed = pyqtSignal()
+
+    def __init__(self, parent: Optional[QObject] = None):
+        """Extended."""
+        super().__init__(parent=parent)
+        # widgets
+        self.label = QLabel("Are you sure to clear?", self)
+        self.buttonBox = QDialogButtonBox(self)
+        self.buttonBox.addButton("OK", QDialogButtonBox.AcceptRole)
+        self.buttonBox.addButton("Cancel", QDialogButtonBox.RejectRole)
+        # connect signals
+        self.buttonBox.accepted.connect(self.buttonOKClicked)
+        self.buttonBox.rejected.connect(self.buttonCancelClicked)
+        # layouts
+        layout = QVBoxLayout(self)
+        layout.addWidget(self.label)
+        layout.addWidget(self.buttonBox)
+
+    def buttonOKClicked(self):
+        """Called when the OK button is clicked."""
+        self.confirmed.emit()
+        self.close()
+
+    def buttonCancelClicked(self):
+        """Called when the Cancel button is clicked."""
+        self.close()
+
+
+class LoggerApp(qiwis.BaseApp):
+    """App for logging.
+
+    Sets a handler of root logger and manages loggerFrame to show log messages.
+    Gives options to clear logs and select log level in the loggerFrame.
+
+    Attributes:
+        loggerFrame: A frame that shows the logs.
+        confirmFrame: A frame that asks whether to clear logs.
+        handler: A handler for adding logs to the loggerFrame. 
+    """
+
+    def __init__(self, name: str, parent: Optional[QObject] = None):
+        """Extended."""
+        super().__init__(name, parent=parent)
+        self.loggerFrame = LoggerFrame()
+        self.confirmFrame = ConfirmClearingFrame()
+        # connect signals to slots
+        self.loggerFrame.clearButton.clicked.connect(self.checkToClear)
+        self.confirmFrame.confirmed.connect(self.clearLog)
+        self.handler = LoggingHandler(self.addLog)
+        # TODO(aijuh): Change the log format when it is determined.
+        fs ="%(levelname)s [%(name)s] [%(filename)s:%(lineno)d] %(message)s "
```

Comment:
Oh, why isn't there a space between `=` and `"`, and why is there a space at the end?
In other words, shouldn't it be `fs = "%(levelname)s ... %(message)s"`?

Reply:
My mistake. I will check details.